tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -187,18 +203,55 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
   }
 
   Status HandleAdd(HloInstruction *instr) override {
-    HloInstruction *bias, *existing_gemm;
+    HloInstruction *bias, *existing_gemm, *slice;
 
-    // First, try to match vector bias add, so we might elide the broadcast.
+    // Attempt to elide broadcast and fuse addition of a vector bias into GEMM.
     if (Match(instr, m::AddAnyOrder(
                          m::Op(&existing_gemm)
                              .WithCustomCallTarget(kCublasLtMatmulCallTarget),
                          m::Broadcast(&bias, m::Op())))) {
       TF_ASSIGN_OR_RETURN(bool was_fused,
-                          FuseVectorBiasAdd(instr, bias, existing_gemm));
+                          FuseVectorBiasAddSliced(instr, bias, existing_gemm));
       if (was_fused) return OkStatus();
     }
 
+    // Attempt to elide broadcast and fuse addition of a vector bias into GEMM
+    // when slicing is applied to the result.
+    if (Match(instr, m::AddAnyOrder(
+                         m::Slice(&slice, m::Op(&existing_gemm)
+                                              .WithCustomCallTarget(
+                                                  kCublasLtMatmulCallTarget)),
+                         m::Broadcast(&bias, m::Op())))) {
+      TF_ASSIGN_OR_RETURN(
+          bool was_fused,
+          FuseVectorBiasAddSliced(instr, bias, existing_gemm, slice));
+      if (was_fused) {
+        return OkStatus();
+      }
+    }
+
+    // Attempt to elide broadcast and fuse addition of a vector bias into
+    // *batched* GEMM as a matrix bias addition using FuseMatrixBiasAdd.
+    // add(bitcast(gemm(a, b)), broadcast(bias)) ->
+    //   bitcast(add(gemm(a, b), bitcast(broadcast(bias)))) ->
+    //   bitcast(gemm(a, b, bitcast(broadcast(bias)))) (FuseMatrixBiasAdd)
+    //
+    if (Match(instr,
+              m::AddAnyOrder(m::Bitcast(m::Op(&existing_gemm)
+                                            .WithCustomCallTarget(
+                                                kCublasLtMatmulCallTarget)),
+                             m::Broadcast(&bias, m::Op())))) {
+      TF_ASSIGN_OR_RETURN(
+          HloInstruction * new_add,
+          MakeBinaryHlo(HloOpcode::kAdd, existing_gemm,
+                        MakeBitcastHlo(bias, existing_gemm->shape())));
+      TF_RETURN_IF_ERROR(
+          ReplaceInstruction(instr, MakeBitcastHlo(new_add, instr->shape())));
```

Comment:
For example here it seems pretty clear that this is incorrect without some WithOneUse() checks?